Plugins/AutocompleteManager/AutocompleteManager.swift

```diff
@@ -0,0 +1,342 @@
+        let newString = (keepPrefixOnCompletion ? session.prefix : "") + autocomplete
+        let newAttributedString = NSAttributedString(string: newString, attributes: attrs)
+        
+        // Modify the NSRange to include the prefix length
+        let rangeModifier = keepPrefixOnCompletion ? session.prefix.count : 0
+        let highlightedRange = NSRange(location: range.location - rangeModifier, length: range.length + rangeModifier)
+        
+        // Replace the attributedText with a modified version including the autocompete
+        let newAttributedText = textView.attributedText.replacingCharacters(in: highlightedRange, with: newAttributedString)
+        if appendSpaceOnCompletion {
+            newAttributedText.append(NSAttributedString(string: " ", attributes: typingTextAttributes))
+        }
+        textView.attributedText = newAttributedText
+    }
+    
+    /// Initializes a session with a new `AutocompleteSession` object
+    ///
+    /// - Parameters:
+    ///   - session: The session to register
+    private func registerCurrentSession(to session: AutocompleteSession) {
+        
+        guard delegate?.autocompleteManager(self, shouldRegister: session.prefix, at: session.range) != false else { return }
+        currentSession = session
+        delegate?.autocompleteManager(self, shouldBecomeVisible: true)
+    }
+    
+    /// Invalidates the `currentSession` session if it existed
+    private func unregisterCurrentSession() {
+        
+        guard let session = currentSession else { return }
+        guard delegate?.autocompleteManager(self, shouldUnregister: session.prefix) != false else { return }
+        currentSession = nil
+        delegate?.autocompleteManager(self, shouldBecomeVisible: false)
+    }
+    
+    /// Calls the required methods to relayout the `AutocompleteTableView` in it's superview
+    private func layoutIfNeeded() {
+        
+        tableView.reloadData()
+        
+        // Resize the table to be fit properly in an `InputStackView`
+        tableView.invalidateIntrinsicContentSize()
+        
+        // Layout the table's superview
+        tableView.superview?.layoutIfNeeded()
+    }
+    
+}
+
+extension AutocompleteManager: UITextViewDelegate {
+    
+    // MARK: - UITextViewDelegate
+    
+    public func textViewDidChange(_ textView: UITextView) {
+        reloadData()
+    }
+    
+    public func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
+        
+        // Ensure that the text to be inserted is not using previous attributes
+        preserveTypingAttributes()
+        
+        // range.length > 0: Backspace/removing text
+        // range.lowerBound < textView.selectedRange.lowerBound: Ignore trying to delete
+        //      the substring if the user is already doing so
+        if range.length > 0, range.lowerBound < textView.selectedRange.lowerBound {
```

Comment:
I think I fixed some bugs in GitHawk related to this logic 🤔 but I'm fine with leaving it for now. we will need unit tests one day

Reply:
This is copied from the githawk fixes